File: src/main/java/com/github/fherbreteau/gatling/ftp/javaapi/protocol/FtpProtocolBuilder.java
package com.github.fherbreteau.gatling.ftp.javaapi.protocol;

import io.gatling.core.protocol.Protocol;
import io.gatling.javaapi.core.ProtocolBuilder;

import java.nio.file.Path;

public class FtpProtocolBuilder implements ProtocolBuilder {

    private final com.github.fherbreteau.gatling.ftp.protocol.FtpProtocolBuilder wrapped;

    public FtpProtocolBuilder(com.github.fherbreteau.gatling.ftp.protocol.FtpProtocolBuilder wrapped) {
        this.wrapped = wrapped;
    }

    public FtpProtocolBuilder server(String server) {
        return new FtpProtocolBuilder(wrapped.server(server));
    }

    public FtpProtocolBuilder port(int port) {
        return new FtpProtocolBuilder(wrapped.port(port));
    }

    public FtpProtocolBuilder credentials(String username, String password) {
        return new FtpProtocolBuilder(wrapped.credentials(username, password));
    }

    public FtpProtocolBuilder localSourcePath(Path sourcePath) {
        return new FtpProtocolBuilder(wrapped.localSourcePath(sourcePath));
    }

    public FtpProtocolBuilder localDestinationPath(Path destpath) {
        return new FtpProtocolBuilder(wrapped.localDestinationPath(destpath));
    }

    public FtpProtocolBuilder remoteSourcePath(Path sourcePath) {
        return new FtpProtocolBuilder(wrapped.remoteSourcePath(sourcePath));
    }

    public FtpProtocolBuilder remoteDestinationPath(Path destpath) {
        return new FtpProtocolBuilder(wrapped.remoteDestinationPath(destpath));
    }

    @Override
    public Protocol protocol() {
        return wrapped.protocol();
    }
}

File: src/main/scala/io/github/fherbreteau/gatling/ftp/protocol/FtpProtocolBuilder.scala
package io.github.fherbreteau.gatling.ftp.protocol

import com.softwaremill.quicklens.ModifyPimp
import io.gatling.commons.model.Credentials
import io.gatling.core.config.GatlingConfiguration

import java.nio.file.Path
import scala.language.implicitConversions

object FtpProtocolBuilder {

  implicit def toFtpProtocol(builder: FtpProtocolBuilder): FtpProtocol = builder.build

  def apply(configuration: GatlingConfiguration): FtpProtocolBuilder = FtpProtocolBuilder(FtpProtocol(configuration))
}

final case class FtpProtocolBuilder(protocol: FtpProtocol) {
  def server(server: String): FtpProtocolBuilder = this.modify(_.protocol.exchange.server).setTo(server)

  def port(port: Int): FtpProtocolBuilder = this.modify(_.protocol.exchange.port).setTo(port)

  def credentials(username: String, password: String): FtpProtocolBuilder = this.modify(_.protocol.exchange.credentials).setTo(Credentials(username, password))

  def localPath(path: Path): FtpProtocolBuilder = this.localSourcePath(path).localDestinationPath(path)

  def localSourcePath(sourcePath: Path): FtpProtocolBuilder = this.modify(_.protocol.localSourcePath).setTo(Some(sourcePath))

  def localDestinationPath(destPath: Path): FtpProtocolBuilder = this.modify(_.protocol.localDestinationPath).setTo(Some(destPath))

  def remotePath(path: String): FtpProtocolBuilder = this.remoteSourcePath(path).remoteDestinationPath(path)

  def remoteSourcePath(sourcePath: String): FtpProtocolBuilder = this.modify(_.protocol.remoteSourcePath).setTo(Some(sourcePath))

  def remoteDestinationPath(destPath: String): FtpProtocolBuilder = this.modify(_.protocol.remoteDestinationPath).setTo(Some(destPath))

  def build: FtpProtocol = protocol
}
